Make allreduce compatible with fx ProxyTensor

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84107
* __->__ #81930
* #83302
* #83303

land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to @Chillee @wanchaol

mrshenli committed:
[WIP] Make allreduce compatible with fx ProxyTensor

[ghstack-poisoned]

kumpera commented on test/distributed/test_c10d_common.py:
```diff
@@ -1236,6 +1237,34 @@ def recv(self, tensor_list, src, tag=0):
+    def test_make_fx(self):
+        os.environ['MASTER_ADDR'] = 'localhost'
+        os.environ['MASTER_PORT'] = '6789'
+        dist.init_process_group("gloo", rank=self.rank, world_size=self.world_size)
```
You must destroy the PG at the end of your test otherwise other tests won't be able to do the same.
You can use a wrapper like this one that simplifies this task:
https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/_shard/sharded_tensor/__init__.py#L81


You need to filter out this test if gloo is not avaible. The way we do it is with a `requires_gloo` wrapper:
https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_distributed.py#L256

kumpera reviewed (changes requested):
Since this test requires process groups, we should run it under both gloo and nccl.

The way we do it is by defining an abstract class with all tests and then on each process group test suite inherit it.

See, CommonDistributedDataParallelTest in test_c10d_common.py and how gloo uses it in test_c10d_gloo.py.

mrshenli committed:
Update on "[WIP] Make allreduce compatible with fx ProxyTensor"




[ghstack-poisoned]

mrshenli commented on test/distributed/test_c10d_common.py:
```diff
@@ -1326,6 +1336,100 @@ def test_send_recv(self):
+        fx_fn = make_fx(fn)(xx)
+
+        y = fn(x)
+        yy = fx_fn(xx).elem
```
@Chillee is there a better way to run the traced `fn` and extract the result?

Chillee replied:
This doesn’t seem right… the traced graph shouldn’t have any notion of proxytensor.

mrshenli replied:
hmm, `fx_fn(xx)` does return a `ProxyTensor`.

mrshenli replied:
oh, it's because I called `t.proxy_mode.tracer.create_proxy` during tracing

mrshenli replied:
it memorized the ProxyTensor during tracing, and use that same tensor instance during execution, instead of correcting the new node to the output from the previous node. Digging

mrshenli committed:
Update on "[WIP] Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38579873).

wanchaol reviewed (commented):
This is nice! Some comments inlined. TBH I am wondering if we should just make `CommTensor` be the default return type of c10d collectives? 

We are anyway moving away from `ProcessGroup:Work` with deprecation message to user, and planning on moving to future. Since future won't get supported well in the compiler, shall we just directly go from `ProcessGroup:Work -> CommTensor`?

mrshenli committed:
Update on "[WIP] Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -18,17 +18,21 @@ c10::intrusive_ptr<ProcessGroup::Work> broadcast_(
           static_cast<ReduceOp>(reduce_op),
           std::chrono::milliseconds(timeout)});
+  return std::
+      tuple<std::vector<at::Tensor>, c10::intrusive_ptr<ProcessGroup::Work>>(
```
That's one possibility, but I didn't make that change in this PR as that might require more discussion. Another option is do that directly in `Ops.cpp`, as  today's implementation in `distributed_c10d.py` won't change the behavior of `c10d::allreduce_` and `c10d::allreduce_` is what get traced.

For this PR, I just wanna add some real tests, so the CommTensor is only added to the test, which is the same as change user program.

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -209,16 +213,19 @@ c10::intrusive_ptr<ProcessGroup::Work> allreduce(
                            int64_t)>();
-  return op.call(
+
+  return std::get<1>(op.call(
```
IIUC, the current change won't have impact on ProcessGroup APIs, as `allreduce_` calls `ProcessGroup::allreduce`, not the other way around. But for the longer term, having a functional version of `ProcessGroup::allreduce` is indeed an option on the table.

aazzolini replied:
Right so this is a new op that got added recently and I think no one is using it so far , correct?

aazzolini replied:
QQ re api, is the returned tensor the same that was passed in the input?

mrshenli replied:
> Right so this is a new op that got added recently and I think no one is using it so far , correct?

Yep, `allreduce_` is recently added and it is not user-facing yet. 

> QQ re api, is the returned tensor the same that was passed in the input?

Yep it is.

mrshenli committed:
Update on "[WIP][Don't Review] Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli commented on test/distributed/test_c10d_common.py:
```diff
@@ -1326,6 +1335,208 @@ def test_send_recv(self):
+                proxy_kwargs = tree_map(unwrap_proxy, kwargs)
+
+                proxy_res = func(*proxy_args, **proxy_kwargs)
+                comm_tensor_proxy = proxy_mode.tracer.create_proxy(
+                    'call_function',
+                    wrap_comm_result,
+                    (proxy_res, ),
+                    {},
+                    name="comm_result"
+                )
+
+                with no_dispatch():
+                    elem_args = tree_map(unwrap_elem, args)
+                    elem_kwargs = tree_map(unwrap_elem, kwargs)
+
+                    out = func(*elem_args, **elem_kwargs)
+
+                return wrap_output(out, comm_tensor_proxy, constant=None, proxy_mode=proxy_mode)
```
Hey @Chillee, below is the `print_tabular` result. Do you know why `aten.mul.Tensor` is still using `add_tensor` instead of `getitem`? 

```
opcode         name                target                                         args                                      kwargs
-------------  ------------------  ---------------------------------------------  ----------------------------------------  --------
placeholder    x_1                 x_1                                            ()                                        {}
call_function  add_tensor          aten.add.Tensor                                (x_1, x_1)                                {}
get_attr       _tensor_constant0   _tensor_constant0                              ()                                        {}
call_function  allreduce__default  c10d.allreduce_.default                        ([add_tensor], _tensor_constant0, 0, -1)  {}
call_function  comm_result         <function wrap_comm_result at 0x7f487f58a820>  (allreduce__default,)                     {}
call_function  getitem             <built-in function getitem>                    (comm_result, 0)                          {}
call_method    call                __call__                                       (getitem, 0)                              {}
call_function  getitem_1           <built-in function getitem>                    (comm_result, 1)                          {}
call_function  mul_tensor          aten.mul.Tensor                                (add_tensor, 2)                           {}
get_attr       _tensor_constant1   _tensor_constant1                              ()                                        {}
output         output              output                                         (_tensor_constant1,)                      {}
``` 

For context, I cannot directly call `func(*args, **kwargs)` for `allreduce_` and operate on the output, because at that time the returned `work` handle no longer has an associated proxy and will be treated as a constant. I believe I have to somehow capture the runtime `work` handle in the graph? So I wanted to directly operate on this [`proxy_res`](https://github.com/pytorch/pytorch/blob/8a6b076196f4169910b2c98d37e1faeaec316050/torch/fx/experimental/proxy_tensor.py#L168) where I still have a chance to use `getitem` to retrieve runtime work handle.

mrshenli replied:
oh I need to update `.proxy` for the input arg as it's inplace

mrshenli committed:
Update on "[WIP][Don't Review] Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP][Don't Review] Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli commented:
two inserted nodes (i.e., wrap comm result + work, and wait on work) are correctly linked in the graph now

```
opcode         name                target                                         args                                      kwargs
-------------  ------------------  ---------------------------------------------  ----------------------------------------  --------
placeholder    x_1                 x_1                                            ()                                        {}
call_function  add_tensor          aten.add.Tensor                                (x_1, x_1)                                {}
get_attr       _tensor_constant0   _tensor_constant0                              ()                                        {}
call_function  allreduce__default  c10d.allreduce_.default                        ([add_tensor], _tensor_constant0, 0, -1)  {}
call_function  comm_result         <function wrap_comm_result at 0x7f2bfd6be8b0>  (allreduce__default,)                     {}                                    
call_function  getitem             <built-in function getitem>                    (comm_result, 0)                          {}                         
call_function  getitem_1           <built-in function getitem>                    (getitem, 0)                              {}                                    
call_function  wait_comm           <function wait_comm at 0x7f2bfd6be790>         (getitem_1,)                              {} 
call_function  mul_tensor          aten.mul.Tensor                                (wait_comm, 2)                            {}                                    
output         output              output                                         (mul_tensor,)                             {}
```

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

aazzolini commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -120,7 +120,6 @@
     ("aten::segment_reduce", datetime.date(2022, 6, 30)),
     ("aten::_segment_reduce_backward", datetime.date(2022, 6, 30)),
     ("aten::empty.SymInt", datetime.date(9999, 1, 1)),
-    ("c10d::broadcast", datetime.date(2022, 6, 25)),
```
Is there a reason to remove this one?

mrshenli replied:
It's already expired back in June. Let me extract this out to a separate PR.

aazzolini commented on test/distributed/test_c10d_common.py:
```diff
@@ -1326,6 +1338,273 @@ def test_send_recv(self):
+    handle. If yes, it will expliclty call wait() in the ``__torch_dispatch__``
+    function before subsequent operations consuming the value of the Tensor.
+
+    In tracing mode, ``CommTensor`` inserts two node into the graph using the
```
```suggestion
    In tracing mode, ``CommTensor`` inserts two nodes into the graph using the
```

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli reviewed (commented):
macos test failures are irrelevant. landing

```
++ python3 .github/scripts/get_workflow_job_id.py 2929184556 i-0b0ee7812ae9e52e2
Traceback (most recent call last):
  File "/Users/ec2-user/runner/_work/pytorch/pytorch/.github/scripts/get_workflow_job_id.py", line 48, in <module>
    jobs = response.json()["jobs"]
  File "/opt/homebrew/lib/python3.9/site-packages/requests/models.py", line 910, in json
    return complexjson.loads(self.text, **kwargs)
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
+ GHA_WORKFLOW_JOB_ID=
Error: Final attempt failed. Child_process exited with error code 1
```

wanchaol reviewed (approved):
Looks great! some minor suggestions inlined. Also I remembered we are switching to modes instead of tensor subclass, is that sth we will do later?

mrshenli commented:
>  Also I remembered we are switching to modes instead of tensor subclass, is that sth we will do later?

The current version is already compatible with the `DispatcherMode` implementation in `proxy_tensor`. The `_ProxyTensor` now is just a dataclass wrapper. 

> I think maybe we can use make_subclass instead of wrapper subclass unless we have specific need? 

Thanks! Let me try that.

wanchaol commented:
> Thanks! Let me try that.
feel free to land this first, we can do this in follow up PRs

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli commented:
@pytorchbot merge -g

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

Differential Revision: [D38579873](https://our.internmc.facebook.com/intern/diff/D38579873)

[ghstack-poisoned]

mrshenli commented:
@pytorchbot merge -g

mrshenli committed:
Update on "Make allreduce compatible with fx ProxyTensor"


land after #83122

This PR explores solutions for 2 issues:

1. Collective comm ops are inplace ops, and does not return a tensor.
   With that, `make_fx` cannot include comm ops in the traced graph.
   The current solution is to make comm ops return a tuple of
   `(output_tensors, work_handle)`, so that
   [`proxy_call`](https://github.com/pytorch/pytorch/blob/90821aab100a436424113e2306eac63f5e247ee5/torch/fx/experimental/proxy_tensor.py#L170-L172)
   can handle that. It won't change the behavior of existing c10d
   Python/C++ APIs, so I directly added the code to `Ops.cpp`.
2. `make_fx` does not recognize `ProcessGroup::Work` and will ignore
   the `wait()` call on the work when tracing graph. However, this
   might break correctness, as when running the traced function, it
   could consume a tensor before it's ready. The current solution
   is to create a `CommTensor` tensor subclass to explicitly call
   `wait()`. In this PR, I am only doing this in the test, as we
   will need more discussion to see if we can add this to c10d Python
   implementations. kudos to Chillee wanchaol

[ghstack-poisoned]

mrshenli commented:
@pytorchbot merge -g

mrshenli commented:
@pytorchbot merge

mrshenli commented:
merged by #84126
